### PR TITLE
fix: stop RefreshTimeout buffer races from returning Internal

### PR DESCRIPTION
**Pull request by grutt**
# Description

Fixes a race in the RefreshTimeout coalesce buffer where empty-buffer singleflight results were returned as gRPC `Internal`, surfacing as nice-grpc `ClientError` on `ctx.refreshTimeout`.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Test changes (add, refactor, improve or change a test)

## What's Changed

- Treat empty RefreshTimeout buffer flushes as no-ops instead of `Internal`
- Re-check pending increments after empty `take`, and force-flush once if a shared no-op stranded the caller's add
- Map finished/missing task runtimes (`pgx.ErrNoRows`) to `NotFound` and drop the buffered increment
- Keep OLAP timeout-refreshed publish best-effort so pub failures don't fail the RPC after a successful extension
- Add a unit test covering empty-take then add/`hasPending`

## Checklist

Changes have been:

- [x] Tested (unit, integration, or manually with steps specified)
- [ ] Linted and formatted
- [ ] Documented (where applicable)
- [ ] Added to CHANGELOG (where applicable) -- see [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## Testing

- `go test ./internal/services/dispatcher/ -run TestRefreshTimeoutBuffer`

---

<details id="ai-disclosure">
<summary><b>🤖 AI Disclosure</b></summary>

- [x] _I acknowledge that an LLM was used in the creation of this Pull Request, in accordance with Hatchet's [AI_POLICY.md](./AI_POLICY.md)._

- **Details**: Investigation via o11y, implementing the buffer race fix and tests, drafting the PR description

</details>

### CODE REVIEW

File: internal/services/dispatcher/server.go [hunk 1]
```diff
@@ -1859,29 +1859,59 @@ func (d *DispatcherImpl) refreshTimeoutV1(ctx context.Context, tenant *sqlcv1.Te
...
+
 	return &contracts.RefreshTimeoutResponse{
-		TimeoutAt: timestamppb.New(timeoutAt.(time.Time)),
+		TimeoutAt: timestamppb.New(timeoutAt),
 	}, nil
 }
 
-func (d *DispatcherImpl) flushRefreshTimeout(ctx context.Context, key refreshTimeoutKey) (time.Time, error) {
-	if !d.refreshTimeoutBuf.shouldFlush(key) {
+func (d *DispatcherImpl) doFlushRefreshTimeout(ctx context.Context, key refreshTimeoutKey, force bool) (time.Time, error) {
```
Comment (mrkaye97) [inline internal/services/dispatcher/server.go:1890]: should we have this return a pointer instead of a zero value? seems silly to check for `time.Zero()` instead of `nil`

File: internal/services/dispatcher/server.go [hunk 2]
```diff
@@ -1859,29 +1859,59 @@ func (d *DispatcherImpl) refreshTimeoutV1(ctx context.Context, tenant *sqlcv1.Te
...
 
+	// A singleflight leader (often a trailing flush) may have observed an empty
+	// buffer before this call's add() was visible. Force-flush once so our
+	// pending increment is not stranded behind a shared no-op result.
+	if timeoutAt.IsZero() && d.refreshTimeoutBuf.hasPending(key) {
+		timeoutAt, err = d.doFlushRefreshTimeout(ctx, key, true)
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (mrkaye97) [inline internal/services/dispatcher/server.go:1875]: I'm not sure I really understand this 🤔 we call flush multiple times here?

File: internal/services/dispatcher/server.go [hunk 3]
```diff
@@ -1900,14 +1930,23 @@ func (d *DispatcherImpl) flushRefreshTimeout(ctx context.Context, key refreshTim
 			if cached, ok := d.refreshTimeoutBuf.lastTimeout(key); ok {
 				return cached, nil
 			}
-			return time.Time{}, fmt.Errorf("no buffered refresh timeout for %s", key)
+			// A singleflight waiter may have add()'d after take saw empty.
+			if d.refreshTimeoutBuf.hasPending(key) {
+				continue
+			}
```
Comment (mrkaye97) [inline internal/services/dispatcher/server.go:1936]: this too, sort of confusing / surprising that either 1) singleflight doesn't protect against this and 2) we don't use a mutex instead to block until the flush completes / make sure this doesn't happen?
